src/nvim/edit.c

```diff
@@ -606,7 +606,10 @@ static int insert_execute(VimState *state, int key)
   s->c = key;
 
   // Don't want K_EVENT with cursorhold for the second key, e.g., after CTRL-V.
-  did_cursorhold = true;
+  // but not in CTRL-X mode, a script can't restore the state //
+  if (ctrl_x_mode == 0) {
+    did_cursorhold = true;
```

Comment:
The other changes in the other files in this commit are fine though. Though they are bizarre, and the Vim patch mentions nothing about the purpose of yet another flag "!".